README.md
# NSString+LevenshteinDistance

[![Version](http://cocoapod-badges.herokuapp.com/v/NSString+LevenshteinDistance/badge.png)](http://cocoadocs.org/docsets/NSString+LevenshteinDistance)
[![Platform](http://cocoapod-badges.herokuapp.com/p/NSString+LevenshteinDistance/badge.png)](http://cocoadocs.org/docsets/NSString+LevenshteinDistance)

## Description

This category adds one method to NSString:

- (NSUInteger)levenshteinDistanceTo:(NSString *)otherString;

## Usage

```objective-c
NSString *a = @"foo@bar.com";
NSString *b = @"foo@bar.con";
NSUInteger distance = [a levenshteinDistanceTo:b];
XCTAssert(distance == 1, @"distance == 1");
```

## Installation

NSString+LevenshteinDistance is available through [CocoaPods](http://cocoapods.org), to install
it simply add the following line to your Podfile:

    pod "NSString+LevenshteinDistance"

## Author

koyachi, rtk2106@gmail.com

## License

NSString+LevenshteinDistance is available under the MIT license. See the LICENSE file for more info.



CHANGELOG.md
# NSString+LevenshteinDistance CHANGELOG

## 0.1.0

Initial release.


Classes/NSString+LevenshteinDistance.m
#import "NSString+LevenshteinDistance.h"

#define min(a, b) ((a <= b) ? a : b)

@implementation NSString (LevenshteinDistance)

- (NSUInteger)levenshteinDistanceTo:(NSString *)otherString
{
    NSUInteger height = self.length + 1;
    NSUInteger width = otherString.length + 1;
    NSMutableArray *distance = [[NSMutableArray alloc] initWithCapacity:width * height];
    NSUInteger      x, y, cost;

    // これいらないかも
    for (y = 0; y < height; y++) {
        for (x = 0; x < width; x++) {
            distance[y * width + x] = @(0);
        }
    }
    for (y = 0; y < height; y++) {
        distance[y * width + 0] = @(y);
    }
    for (x = 0; x < width; x++) {
        distance[x] = @(x);
    }

    for (y = 1; y < height; y++) {
        for (x = 1; x < width; x++) {
            unichar c1 = [self characterAtIndex:y - 1];
            unichar c2 = [otherString characterAtIndex:x - 1];
            cost = c1 == c2 ? 0 : 1;

            NSUInteger insert  = ((NSNumber *) distance[(y - 1) * width + x]).integerValue + 1;
            NSUInteger remove  = ((NSNumber *) distance[y * width + x - 1]).integerValue + 1;
            NSUInteger replace = ((NSNumber *) distance[(y - 1) * width + x - 1]).integerValue + cost;
            distance[y * width + x] = @(min(min(insert, remove), replace) );
//            NSLog(@"d(%d,%d) = %@", y, x, distance[y * width + x]);
        }
    }

    return ((NSNumber *) distance[height * width - 1]).integerValue;
}

@end


Example/NSString+LevenshteinDistanceDemo/AppDelegate.m
//
//  AppDelegate.m
//  NSString+LevenshteinDistanceDemo
//
//  Created by koyachi on 2014/04/13.
//  Copyright (c) 2014年 koyachi. All rights reserved.
//

#import "AppDelegate.h"

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    // Override point for customization after application launch.
    return YES;
}
							
- (void)applicationWillResignActive:(UIApplication *)application
{
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application
{
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later. 
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
    // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application
{
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application
{
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

@end


Example/NSString+LevenshteinDistanceDemo/ViewController.m
//
//  ViewController.m
//  NSString+LevenshteinDistanceDemo
//
//  Created by koyachi on 2014/04/13.
//  Copyright (c) 2014年 koyachi. All rights reserved.
//

#import "ViewController.h"
#import "NSString+LevenshteinDistance.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    self.dstTextField.placeholder = @"foo";
    self.dstTextField.delegate = self;
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (BOOL)textField:(UITextField *)textiField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string
{
    if ([string isEqualToString:@"\n"]) {
        NSUInteger distance = [self.srcLabel.text levenshteinDistanceTo:self.dstTextField.text];
        self.distanceValueLabel.text = @(distance).stringValue;
        NSLog(@"distance = %d", distance);
    }

    return YES;
}

@end


Example/NSString+LevenshteinDistanceDemoTests/NSString_LevenshteinDistanceDemoTests.m
//
//  NSString_LevenshteinDistanceDemoTests.m
//  NSString+LevenshteinDistanceDemoTests
//
//  Created by koyachi on 2014/04/13.
//  Copyright (c) 2014年 koyachi. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "NSString+LevenshteinDistance.h"

@interface NSString_LevenshteinDistanceDemoTests : XCTestCase

@end

@implementation NSString_LevenshteinDistanceDemoTests

- (void)setUp
{
    [super setUp];
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown
{
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

- (void)testExample
{
    NSString *a = @"foo@bar.com";
    NSString *b = @"foo@bar.con";
    NSUInteger distance = [a levenshteinDistanceTo:b];
    NSLog(@"distance = %d", distance);
    XCTAssert(distance == 1, @"distance == 1");

    a = @"foo@bar.com";
    b = @"foo@bar.cmo";
    distance = [a levenshteinDistanceTo:b];
    NSLog(@"distance = %d", distance);
    XCTAssert(distance == 2, @"distance == 2");
}

@end
